src/backends/matrix_state_interleaved.hpp

```diff
@@ -0,0 +1,248 @@
+#pragma once
+
+#include <memory/memory.hpp>
+#include <util/span.hpp>
+#include <util/rangeutil.hpp>
+
+#include "gpu_kernels.hpp"
+
+namespace nest {
+namespace mc {
+namespace gpu {
+
+/// matrix state
+template <typename T, typename I>
+struct matrix_state_interleaved {
+    using value_type = T;
+    using size_type = I;
+
+    using array  = memory::device_vector<value_type>;
+    using iarray = memory::device_vector<size_type>;
+
+    using const_view = typename array::const_view_type;
+
+    //
+    // Permutation and index information required for forward and backward
+    // interleave-permutation of vectors.
+    //
+
+    // size of each matrix (after permutation in ascending size)
+    iarray matrix_sizes;
+    // start values corresponding to matrix i in the external storage
+    iarray matrix_index;
+
+    //
+    // Storage for the matrix and parent index in interleaved format.
+    // Includes the cv_capacitance, which is required for matrix assembly.
+    //
+    iarray parent_index;
+    array d;   // [μS]
+    array u;   // [μS]
+    array rhs; // [nA]
+
+    // required for matrix assembly
+    array cv_capacitance; // [pF]
+
+    // the invariant part of the matrix diagonal
+    array invariant_d;    // [μS]
+
+    // the length of a vector required to store values for one
+    // matrix with padding
+    int padded_size;
+
+    //  Storage for solution in uninterleaved format.
+    //  Used to hold the storage for passing to caller, and must be updated
+    //  after each call to the ::solve() method.
+    array solution;
+
+    // default constructor
+    matrix_state_interleaved() = default;
+
+    // Construct matrix state for a set of matrices defined by parent_index p
+    // The matrix solver stores the matrix in an "interleaved" structure for
+    // optimal solution, which requires a significant amount of precomputing
+    // of indexes and data structures in the constructor.
+    //  cv_cap      // [pF]
+    //  face_cond   // [μS]
+    matrix_state_interleaved(const std::vector<size_type>& p,
+                 const std::vector<size_type>& cell_index,
+                 const std::vector<value_type>& cv_cap,
+                 const std::vector<value_type>& face_cond)
+    {
+        // Assert that capacitance and conductance satisfie one of two conditions
+        // * both are defined, with one value for each compartment
+        // * both are empty
+        // Note that the number of compartments is equal to p.size()
+        EXPECTS(   cv_cap.size()==face_cond.size()
```

Comment:
I think it is now safe to require that `cv_cap` et al. be filled out when the constructor is called. As mentioned above, the specification for a matrix state implementation class should be made explicitly in a comment or document.